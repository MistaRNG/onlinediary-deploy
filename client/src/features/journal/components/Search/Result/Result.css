.search-result {
  overflow-x: hidden;
  margin: 0.8em 0;
  flex: 1;
  overflow-y: auto;
  max-height: 40vh;
}

.search-result > * {
  margin: 1em 0.2em;
  max-width: 100%;
}

.result-list-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: relative;
}

.result-content {
  flex: 1;
}

.delete-button {
  position: absolute;
  right: 10px;
  top: 50%;
  transform: translateY(-50%);
  background-color: transparent;
  border: none;
  cursor: pointer;
}
