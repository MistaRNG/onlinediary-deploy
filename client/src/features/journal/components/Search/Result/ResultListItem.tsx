import { Card, Elevation, Button, Icon } from "@blueprintjs/core";
import { useNavigate } from "react-router-dom";
import { getResultDate } from "common/helpers";
import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { deleteJournal } from "features/journal/journalSlice";
import { AppDispatch } from "app/store";

interface ResultListItemProps {
  date: string;
  title?: string;
}

const ResultListItem: React.FC<ResultListItemProps> = ({ date, title }) => {
  const navigate = useNavigate();
  const dispatch: AppDispatch = useDispatch();

  const [hovered, setHovered] = useState(false);

  const clickHandler = () => navigate(`/journal/${date}`);

  const deleteHandler = (event: React.MouseEvent) => {
    event.stopPropagation();

    const confirmed = window.confirm("Are you sure you want to delete this journal entry?");
    if (confirmed) {
      dispatch(deleteJournal(date));
    }
  };

  return (
    <Card onClick={clickHandler} interactive={true} elevation={Elevation.TWO} className="result-list-item">
      <div className="result-content">
        <h5>{getResultDate(date)}</h5>
        <p>{title || "Untitled"}</p>
      </div>
      <Button 
        icon={<Icon icon="trash" style={{ color: hovered ? '#fb3030' : '#fd5d5d' }} />} 
        minimal={true} 
        onClick={deleteHandler} 
        className="delete-button"
        onMouseEnter={() => setHovered(true)}
        onMouseLeave={() => setHovered(false)}
      />
    </Card>
  );
};

export default ResultListItem;
