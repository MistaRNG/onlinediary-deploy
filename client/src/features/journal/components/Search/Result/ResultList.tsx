import "./Result.css";
import ResultListItem from "./ResultListItem";
import React from "react";

interface ResultProps {
  results: Array<{
    id: string;
    title: string;
    date: string;
  }>;
}

const Result: React.FC<ResultProps> = ({ results }) => {
  const resultElms = results.map((info, i) => (
    <ResultListItem {...info} key={i} />
  ));

  return (
    <div className="search-result">
      {results.length ? resultElms : "No journal is found"}
    </div>
  );
};

export default Result;
