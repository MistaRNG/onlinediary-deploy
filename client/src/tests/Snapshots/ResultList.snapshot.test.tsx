import React from 'react';
import { render } from '@testing-library/react';
import Result from '../../features/journal/components/Search/Result/ResultList';
import '@testing-library/jest-dom';
import { describe, it, expect } from '@jest/globals';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';

const mockReducer = (state = {}, action: any) => state;

const mockStore = configureStore({
  reducer: {
    mock: mockReducer,
  },
});

describe('Result Component Snapshot Tests', () => {
  it('should render the Result component correctly with results', () => {
    // ARRANGE
    const mockResults = [
      { id: '1', title: 'First Journal', date: '2024-09-06' },
      { id: '2', title: 'Second Journal', date: '2024-09-07' },
    ];

    // ACT
    const { asFragment } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <Result results={mockResults} />
        </BrowserRouter>
      </Provider>
    );

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the Result component correctly with no results', () => {
    // ARRANGE
    const mockResults: Array<{ id: string; title: string; date: string }> = [];

    // ACT
    const { asFragment } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <Result results={mockResults} />
        </BrowserRouter>
      </Provider>
    );

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });
});
