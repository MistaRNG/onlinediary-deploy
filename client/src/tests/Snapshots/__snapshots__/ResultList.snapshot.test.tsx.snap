// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Component Snapshot Tests should render the Result component correctly with no results 1`] = `
<DocumentFragment>
  <div
    class="search-result"
  >
    No journal is found
  </div>
</DocumentFragment>
`;

exports[`Result Component Snapshot Tests should render the Result component correctly with results 1`] = `
<DocumentFragment>
  <div
    class="search-result"
  >
    <div
      class="bp4-card bp4-interactive bp4-elevation-2 result-list-item"
    >
      <div
        class="result-content"
      >
        <h5>
          September 6, 2024
        </h5>
        <p>
          First Journal
        </p>
      </div>
      <button
        class="bp4-button bp4-minimal delete-button"
        type="button"
      >
        <span
          aria-hidden="true"
          class="bp4-icon bp4-icon-trash"
          icon="trash"
          style="color: rgb(253, 93, 93);"
        >
          <svg
            data-icon="trash"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M14.49 3.99h-13c-.28 0-.5.22-.5.5s.22.5.5.5h.5v10c0 .55.45 1 1 1h10c.55 0 1-.45 1-1v-10h.5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5zm-8.5 9c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm2-12h-4c0-.55-.45-1-1-1h-2c-.55 0-1 .45-1 1h-4c-.55 0-1 .45-1 1v1h14v-1c0-.55-.45-1-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="bp4-card bp4-interactive bp4-elevation-2 result-list-item"
    >
      <div
        class="result-content"
      >
        <h5>
          September 7, 2024
        </h5>
        <p>
          Second Journal
        </p>
      </div>
      <button
        class="bp4-button bp4-minimal delete-button"
        type="button"
      >
        <span
          aria-hidden="true"
          class="bp4-icon bp4-icon-trash"
          icon="trash"
          style="color: rgb(253, 93, 93);"
        >
          <svg
            data-icon="trash"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M14.49 3.99h-13c-.28 0-.5.22-.5.5s.22.5.5.5h.5v10c0 .55.45 1 1 1h10c.55 0 1-.45 1-1v-10h.5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5zm-8.5 9c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm2-12h-4c0-.55-.45-1-1-1h-2c-.55 0-1 .45-1 1h-4c-.55 0-1 .45-1 1v1h14v-1c0-.55-.45-1-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
